expr/func_aspercent.go

```diff
@@ -0,0 +1,241 @@
+package expr
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/grafana/metrictank/api/models"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+type FuncAsPercent struct {
+	in          GraphiteFunc
+	totalFloat  float64
+	totalSeries GraphiteFunc
+	nodes       []expr
+}
+
+func NewAsPercent() GraphiteFunc {
+	return &FuncAsPercent{totalFloat: math.NaN()}
+}
+
+func (s *FuncAsPercent) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgIn{
+			key: "total",
+			opt: true,
+			args: []Arg{
+				ArgFloat{val: &s.totalFloat},
+				ArgSeriesList{val: &s.totalSeries},
+			},
+		},
+		ArgStringsOrInts{val: &s.nodes, opt: true, key: "nodes"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncAsPercent) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncAsPercent) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var outSeries []models.Series
```

Comment:
Could you do a 

```
defer cache[Req{}] = append(cache[Req{}], outSeries...)
```

here instead of all the individual cache steps?

Reply:
very nice, I like it

Reply:
defer args are evaluated at call time, so this won't work as expected.
see https://play.golang.org/p/06o0zrVeOH-

Reply:
You forgot to call foo() in your snippet.
see https://play.golang.org/p/Qh4rp1p2q73